### DISCUSSION TITLE
The add-on backup could be smaller if I ignore the `node_modules` folder, is it possible?

### DISCUSSION THREAD

**Discussion by dougiteixeira** [hassio-addons/addon-node-red#1910] [General] [answered]
The backup of the Node-RED add-on is averaging 19MB, while the backup of other add-ons is less than 0.1MB.

I investigated and the `addon_configs\a0d7b954_nodered\node_modules` folder is being included in the backup, in my case it is 55MB before being compressed.

I also noticed that the lack of this folder does not harm the addon, as it recreates the folder and contents when it starts.

So the question is: Is it really necessary to include this folder in the add-on backup?

**Comment by frenck** [comment]
Hmm good point!

**Comment by frenck** [comment, marked as answer]
https://github.com/hassio-addons/addon-node-red/pull/1957

**Reply by finipini** [reply]
> #1957

That pr is presented has a new option on the configuration.yaml folder but there is nothing about that option on the documentation.

Is it any way of using it to not exclude the modules from the backup for the ones who liked to stay as it is ? 


![Screenshot_2024-10-05-10-44-32-251_io homeassistant companion android-edit](https://github.com/user-attachments/assets/444fa012-7da9-469c-8cc5-2e80628d590d)


Like this 👆 , for example .

**Reply by frenck** [reply]
It doesn't.

> Is it any way of using it to not exclude the módulos from the backup for the ones whonliked to stay as it is ?

Nope.

**Comment by 3ative** [comment]
I'm not so sure it's a good idea. 
* Rename node_modules >> node_modules_OLD 
* Restarting NR so node_modules is remade, causes loads of errors:
```
npm notice
npm notice New minor version of npm available! 10.7.0 -> 10.8.3
npm notice Changelog: https://github.com/npm/cli/releases/tag/v10.8.3
npm notice To update run: npm install -g npm@10.8.3
npm notice
npm error code 1
npm error path /config/node_modules/websocket
npm error command failed
npm error command sh -c node-gyp rebuild
npm error gyp info it worked if it ends with ok
npm error gyp info using node-gyp@10.1.0
npm error gyp info using node@18.20.3 | linux | x64
npm error gyp ERR! find Python 
npm error gyp ERR! find Python Python is not set from command line or npm configuration
npm error gyp ERR! find Python Python is not set from environment variable PYTHON
npm error gyp ERR! find Python checking if "python3" can be used
npm error gyp ERR! find Python - executable path is ""
npm error gyp ERR! find Python - "" could not be run
npm error gyp ERR! find Python checking if "python" can be used
npm error gyp ERR! find Python - executable path is ""
npm error gyp ERR! find Python - "" could not be run
npm error gyp ERR! find Python 
npm error gyp ERR! find Python **********************************************************
npm error gyp ERR! find Python You need to install the latest version of Python.
npm error gyp ERR! find Python Node-gyp should be able to find and use Python. If not,
npm error gyp ERR! find Python you can try one of the following options:
npm error gyp ERR! find Python - Use the switch --python="/path/to/pythonexecutable"
npm error gyp ERR! find Python (accepted by both node-gyp and npm)
npm error gyp ERR! find Python - Set the environment variable PYTHON
npm error gyp ERR! find Python - Set the npm configuration variable python:
npm error gyp ERR! find Python npm config set python "/path/to/pythonexecutable"
npm error gyp ERR! find Python For more information consult the documentation at:
npm error gyp ERR! find Python https://github.com/nodejs/node-gyp#installation
npm error gyp ERR! find Python **********************************************************
npm error gyp ERR! find Python 
npm error gyp ERR! configure error 
npm error gyp ERR! stack Error: Could not find any Python installation to use
npm error gyp ERR! stack at PythonFinder.fail (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:306:11)
npm error gyp ERR! stack at PythonFinder.findPython (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:164:17)
npm error gyp ERR! stack at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
npm error gyp ERR! stack at async configure (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:27:18)
npm error gyp ERR! stack at async run (/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js:81:18)
npm error gyp ERR! System Linux 6.6.46-haos
npm error gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm error gyp ERR! cwd /config/node_modules/websocket
npm error gyp ERR! node -v v18.20.3
npm error gyp ERR! node-gyp -v v10.1.0
npm error gyp ERR! not ok

npm error A complete log of this run can be found in: /root/.npm/_logs/2024-09-30T21_14_25_445Z-debug-0.log
[22:14:49] WARNING: Failed un-installing conflicting packages
s6-rc: info: service init-nodered successfully started
s6-rc: info: service nodered: starting
s6-rc: info: service nodered successfully started
s6-rc: info: service nginx: starting
s6-rc: info: service nginx successfully started
s6-rc: info: service legacy-services: starting
[22:14:50] INFO: Starting Node-RED...
s6-rc: info: service legacy-services successfully started
```
And within NR loads of flows fail to start due to missing "node types"

**Reply by frenck** [reply]
That means you are not using the documented options for installing additional nodes and doing it manually. 🤷‍♂️

**Reply by 3ative** [reply]
> That means you are not using the documented options for installing additional nodes and doing it manually. 🤷‍♂️

I can't be the only one, so this change will break others install too.
Also, I see nothing in the DOCs (Add-ons>>Node-RED>>Documentation tab) about "installing additional nodes"
- Is there somewhere else I should be looking?

**Comment by 3ative** [comment]
Ok, I found this - After much searching for: "installing additional nodes". 🤦‍♂️

![doc-ish](https://github.com/user-attachments/assets/89de2398-17b8-4244-8ff3-69d590b304c6)

So the workflow now MUST be:
1. Open Node-Red, go to "Manage Palettes" > "Install" and use "search modules" to find the Extra Node(s) we want. 
2. Make a note of name(s)
3. Go back to Add-ons>Node-RED, "Configuration Tab"
4. Add name(s) to "npm_packages" or edit the YAML
5. Restart NR
6. Back to NR>"Manage Palettes" to see if it worked/installed

**Reply by jmav** [reply]
Agree with proper approach. This should be warned in upgrade and add a guidance to fix it - a lot of people have an issue with this. This is huge change.

**Reply by frenck** [reply]
> This is huge change.

It is not, this option has been there since day 1 this add-on was published.

**Reply by jmav** [reply]
I mean removing node_modules (agree with action) can have huge effect on end user, preparing good guidance could help.

**Comment by 3ative** [comment]
@jmav I went though everything and made a quick explainer and guide on what should be done and how to change over the old (Editor installed) Palettes.
I hope this helps:
[Node Red Update v18.1.1 Old Vs Proper Palette installs](https://youtu.be/A87elZ5Z-PQ)

**Comment by 3ative** [comment]
This update also seems to have broken templating?
https://github.com/hassio-addons/addon-node-red/issues/1968

**Reply by frenck** [reply]
nope

**Reply by 3ative** [reply]
A few have noted templating stopped working and it only happened after this update.

**Reply by frenck** [reply]
🤷

**Reply by 3ative** [reply]
🤣

**Reply by 3ative** [reply]
If you get time, I'd appreciate so pointers on how to diagnose (Links, Etc). I'd be more-than happy to do the work, I just know enough to be useful atm. 🙏

**Reply by frenck** [reply]
Not sure how that is related to the original OP here at all. So... pretty much approaching this as noise tbh.

**Reply by 3ative** [reply]
I hear ya.... It may be a coincidence. Seems a bit too "on the nose" thou. :)

**Reply by 3ative** [reply]
I found that "Render Template" Node still works, but "{{states.sensor.bedroom_temperature.state}}" doesn't get "rendered" anywhere else.
![TP-Node-Tester](https://github.com/user-attachments/assets/1d2493a9-570b-44a9-a9bd-cb097f1c37a4)

**Reply by frenck** [reply]
allright fully off topic, locking this down.

Additionally, the template uses an invalid statement:

`{{states.sensor.bedroom_temperature.state}}`

You should not be using `states` directly that way.
